vm/src/vm/runners/cairo_pie.rs

```diff
@@ -272,61 +393,117 @@ mod serde_impl {
+            };
+
+            let mut res = vec![];
+            for cell_bytes in bytes.chunks(CELL_BYTE_LEN) {
+                let addr = relocatable_from_bytes(cell_bytes[0..ADDR_BYTE_LEN].try_into().ok()?);
+                let field_bytes = &cell_bytes[ADDR_BYTE_LEN..CELL_BYTE_LEN];
+                // Check the last bit to determine if it is a Relocatable or Felt value
+                let value = if (field_bytes[field_bytes.len() - 1] & 0x80) != 0 {
+                    let (segment, offset) =
+                        relocatable_from_bytes(field_bytes[0..ADDR_BYTE_LEN].try_into().ok()?);
+                    MaybeRelocatable::from((segment as isize, offset))
+                } else {
+                    MaybeRelocatable::from(Felt252::from_bytes_le_slice(field_bytes))
+                };
+                res.push((addr, value));
+            }
+
+            Some(CairoPieMemory(res))
+        }
     }
 
-    pub fn serialize_signature_additional_data<S>(
-        values: &HashMap<Relocatable, (Felt252, Felt252)>,
-        serializer: S,
-    ) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let mut seq_serializer = serializer.serialize_seq(Some(values.len()))?;
-
-        for (key, (x, y)) in values {
-            seq_serializer.serialize_element(&[
-                [
-                    Felt252Wrapper(&Felt252::from(key.segment_index)),
-                    Felt252Wrapper(&Felt252::from(key.offset)),
-                ],
-                [Felt252Wrapper(x), Felt252Wrapper(y)],
-            ])?;
+    pub mod signature_additional_data {
+        use super::*;
+
+        pub fn serialize<S>(
+            values: &HashMap<Relocatable, (Felt252, Felt252)>,
+            serializer: S,
+        ) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            let mut seq_serializer = serializer.serialize_seq(Some(values.len()))?;
+
+            for (key, (x, y)) in values {
+                seq_serializer.serialize_element(&[
+                    [
+                        Felt252Wrapper(&Felt252::from(key.segment_index)),
+                        Felt252Wrapper(&Felt252::from(key.offset)),
+                    ],
+                    [Felt252Wrapper(x), Felt252Wrapper(y)],
+                ])?;
+            }
+            seq_serializer.end()
+        }
+
+        pub fn deserialize<'de, D>(
+            d: D,
+        ) -> Result<HashMap<Relocatable, (Felt252, Felt252)>, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            let number_map = Vec::<((Number, Number), (Number, Number))>::deserialize(d)?;
+            let mut res = HashMap::new();
+            for ((index, offset), (r, s)) in number_map.into_iter() {
+                let addr = Relocatable::from((
+                    index.as_u64().ok_or(D::Error::custom("Invalid address"))? as isize,
+                    offset.as_u64().ok_or(D::Error::custom("Invalid address"))? as usize,
```

Comment:
Are these errors allocated on the heap? Maybe use `ok_or_else` so we only pay for them when actually needed?